src/data_structures/array_2d.rs

```diff
@@ -7,10 +7,13 @@ pub struct Array2D<T> {
     height: usize,
 }
 
-impl<T> Array2D<T> {
+impl<T> Array2D<T>
+where
+    T: Default + Clone,
+{
     pub fn new(width: usize, height: usize) -> Self {
         Array2D {
-            array: Vec::with_capacity(width * height),
+            array: vec![T::default(); width * height],
```

Comment:
having a `default` method convention where types declare a default value is a great idea.
Looks like it's a trait
https://doc.rust-lang.org/std/default/trait.Default.html